builtin/common/serialize.lua

```diff
@@ -1,205 +1,190 @@
+    return counts
+end
 
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
-
-	-- Mutually recursive local functions:
-	local dump_val, dump_or_ref_val
+-- Build a table with the succeeding character from A-Z
+local succ = {}
+for char = ("A"):byte(), ("Z"):byte() - 1 do
+    succ[string.char(char)] = string.char(char + 1)
+end
 
-	-- If x occurs multiple times, dump the local variable rather than
-	-- the value. If it's the first time it's dumped, also dump the
-	-- content in local_defs.
-	function dump_or_ref_val(x)
-		if seen[x] ~= 2 then
-			return dump_val(x)
-		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
-		end
-		-- First occurence, create and register reference
-		local val = dump_val(x)
-		local i = local_index
-		local_index = local_index + 1
-		var = "_["..i.."]"
-		local_defs[#local_defs + 1] = var.." = "..val
-		dumped[x] = var
-		return var
-	end
+local function quote(string)
+    return string_format("%q", string)
+end
 
-	-- Second phase.  Dump the object; subparts occuring multiple times
-	-- are dumped in local variables which can be referenced multiple
-	-- times.  Care is taken to dump local vars in a sensible order.
-	function dump_val(x)
-		local  tp = type(x)
-		if     x  == nil        then return "nil"
-		elseif tp == "string"   then return string.format("%q", x)
-		elseif tp == "boolean"  then return x and "true" or "false"
-		elseif tp == "function" then
-			return string.format("loadstring(%q)", string.dump(x))
-		elseif tp == "number"   then
-			-- Serialize numbers reversibly with string.format
-			return string.format("%.17g", x)
-		elseif tp == "table" then
-			local vals = {}
-			local idx_dumped = {}
-			local np = nest_points[x]
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
-				end
-				idx_dumped[i] = true
-			end
-			for k, v in pairs(x) do
-				if (not np or not np[k]) and
-						not idx_dumped[k] then
-					vals[#vals + 1] = "["..dump_or_ref_val(k).."] = "
-						..dump_or_ref_val(v)
-				end
-			end
-			return "{"..table.concat(vals, ", ").."}"
-		else
-			error("Can't serialize data of type "..tp)
+local function serialize(value, write)
+    local reference = {"A"}
+    local function increment_reference(place)
+        if not reference[place] then
+            reference[place] = "B"
+        elseif reference[place] == "Z" then
+            reference[place] = "A"
+            return increment_reference(place + 1)
+        else
+            reference[place] = succ[reference[place]]
+        end
+    end
+    local references = {}
+    local to_fill = {}
+    for object, count in pairs(count_objects(value)) do
+        local type_ = type(object)
+        if count >= 2 and (type_ ~= "string" or #reference + 2 < #object) then
+            local ref = table_concat(reference)
+            write(ref)
+            write"="
```

Comment:
Inconsistent use of parentheses. Whereas this is a valid notation, it does make the code less readable.

Reply:
Will adhere